traversal/planner/GraphPlanner.java

```diff
@@ -231,37 +233,43 @@ Optimiser optimiser() {
     }
 
     void mayOptimise(GraphManager graphMgr, boolean singleUse) {
-        if (procedure == null) {
-            try {
-                firstOptimisingLock.writeLock().lock();
-                if (procedure == null) optimise(graphMgr, singleUse);
-                assert procedure != null;
-            } finally {
-                firstOptimisingLock.writeLock().unlock();
-            }
-        } else if (isOptimising.compareAndSet(false, true)) {
-            try {
-                optimise(graphMgr, singleUse);
-            } finally {
-                isOptimising.set(false);
-            }
+        long timeLimitMillis = singleUse ? HIGHER_TIME_LIMIT_MILLIS : DEFAULT_TIME_LIMIT_MILLIS;
+        if (backgroundOptimisation == null) beginFirstOptimise(graphMgr, timeLimitMillis);
+        else if (isOptimising.compareAndSet(false, true)) beginReOptimise(graphMgr, timeLimitMillis);
```

Comment:
if we have not started any optimisation yet, we start the first optimisation. Otherwise, only 1 thread is allowed to go and reset the future to run another optimisation.

The `isOptimising` AtomicBoolean is set back to `false` within the background optimisation task.